## Hotel Dataset — 30 Data Cleaning Questions

#### 1. Remove leading/trailing spaces from column names
#### Explanation: Column names like "Booking ID " and "Customer Name" have extra spaces. This
#### causes errors when you try to select columns. Just strip the spaces so names become clean
#### (Booking ID, Customer Name).

In [1]:
import pandas as pd

df = pd.read_csv("C:\\Users\\rohit\\Downloads\\Hotel_booking_dataset.csv")

df.columns = df.columns.str.strip()

print(df.columns)

Index(['Booking ID', 'Customer Name', 'Booking Date', 'City', 'Room Type',
       'Booking Channel', 'Customer Segment', 'Gender', 'Age', 'Room Nights',
       'Room Rate', 'Discount %', 'Rating', 'Guests', 'Is Member', 'Cancelled',
       'Payment Method', 'Gross Amount', 'Discount Amount', 'Final Amount',
       'Hotel Cost', 'Profit'],
      dtype='object')


#### 2. Standardize all date formats into one consistent format (YYYY-MM-DD)
#### Explanation: Dates appear as 2025-05-17 00:00:00, 2025/02/20, 15/01/2025, and even
#### invalid ones like 32/13/2025. Convert everything to a single clean format and turn bad dates
#### into missing values.

In [2]:
df["Booking Date"] = pd.to_datetime(
    df["Booking Date"],
    errors="coerce"
)

df["Booking Date"] = df["Booking Date"].dt.strftime("%Y-%m-%d")

df["Booking Date"]

0       2025-05-17
1       2025-07-29
2       2025-10-19
3       2025-12-14
4       2025-01-20
           ...    
2020    2025-11-27
2021    2025-10-22
2022    2025-07-11
2023    2025-01-24
2024    2025-04-06
Name: Booking Date, Length: 2025, dtype: object

#### 3. Handle completely missing Customer Name values
#### Explanation: Some rows have blank Customer Name or values like N/A. Decide whether to
#### fill them with a placeholder (e.g., “Unknown”) or drop those rows, depending on whether
#### the name is critical for analysis.

In [3]:
df["Customer Name"] = df["Customer Name"].replace(
    ["", "N/A", "NA", "n/a", "None"],
    pd.NA
)

df["Customer Name"] = df["Customer Name"].fillna("Unknown")

print(df["Customer Name"].value_counts(dropna=False))

Customer Name
Unknown          34
john doe          6
12345             5
unknown           4
 JOHN DOE         4
                 ..
Customer_1632     1
Customer_0725     1
Customer_0041     1
Customer_0658     1
Customer_0642     1
Name: count, Length: 1951, dtype: int64


#### 4. Fix inconsistent Gender values
#### Explanation: Gender appears as Male, FEMALE, male, F, M, Other, etc. Convert everything to
#### a standard set such as Male, Female, Other.

In [4]:
df["Gender"] = (
    df["Gender"]
    .astype("string")
    .str.strip()
    .str.lower()
)

gender_map = {
    "male": "Male",
    "m": "Male",
    "female": "Female",
    "f": "Female",
    "other": "Other"
}

df["Gender"] = df["Gender"].map(gender_map)

print(df["Gender"].value_counts(dropna=False))

Gender
Female    1004
Male       952
Other       69
Name: count, dtype: int64


#### 5. Clean and standardize City names
#### Explanation: Cities have typos and casing issues (mumbai, Mumbai , Bangalore, Hydrabad,
#### blank). Correct spelling, remove extra spaces, and make casing consistent (e.g., Title Case).

In [5]:
df["City"] = (
    df["City"]
    .astype("string")
    .str.strip()
    .str.lower()
)

city_map = {
    "mumbai": "Mumbai",
    "bangalore": "Bengaluru",
    "bengaluru": "Bengaluru",
    "hydrabad": "Hyderabad",
    "hyderabad": "Hyderabad"
}

df["City"] = df["City"].map(city_map)

df["City"] = df["City"].fillna("Unknown")

print(df["City"].value_counts())

City
Unknown      1036
Mumbai        358
Bengaluru     324
Hyderabad     307
Name: count, dtype: int64


#### 6. Standardize Room Type categories
#### Explanation: Room Type has Suite, Deluxe, delux, standard, exec, Unknown, and blanks.
#### Map similar values to a clean list: Standard, Deluxe, Suite, Executive.

In [6]:
df["Room Type"] = (
    df["Room Type"]
    .astype("string")
    .str.strip()
    .str.lower()
)

room_type_map = {
    "standard": "Standard",
    "deluxe": "Deluxe",
    "delux": "Deluxe",
    "suite": "Suite",
    "executive": "Executive",
    "exec": "Executive"
}

df["Room Type"] = df["Room Type"].map(room_type_map)

df["Room Type"] = df["Room Type"].fillna("Unknown")

print(df["Room Type"].value_counts())

Room Type
Deluxe       548
Executive    506
Suite        479
Standard     454
Unknown       38
Name: count, dtype: int64


#### 7. Clean Booking Channel values
#### Explanation: Values include Online, online, Walk-in, Travel Agent, TravelAgent, corp,
#### Corporate. Unify them into a small consistent set (Online, Walk-in, Corporate, Travel Agent).

In [7]:
df['Booking Channel'] = (
    df['Booking Channel']
    .astype(str)
    .str.strip()
    .str.lower()
)

df['Booking Channel'] = df['Booking Channel'].replace({
    'online': 'Online',
    'walk-in': 'Walk-in',
    'walk in': 'Walk-in',
    'travel agent': 'Travel Agent',
    'travelagent': 'Travel Agent',
    'corp': 'Corporate',
    'corporate': 'Corporate'
})

df['Booking Channel'] = df['Booking Channel'].replace(
    ['', 'nan', 'unknown', 'n/a', 'na'],
    pd.NA
)

df['Booking Channel'].value_counts(dropna=False)

Booking Channel
Online          555
Walk-in         527
Travel Agent    472
Corporate       467
walkin            4
Name: count, dtype: int64

#### 8. Standardize Customer Segment
#### Explanation: Segments appear as Premium, regular, REG, Budget , VIP, etc. Create a clean set
#### such as Budget, Regular, Premium, Corporate, VIP and map everything to it.

In [8]:
df['Customer Segment'] = (
    df['Customer Segment']
    .astype(str)
    .str.strip()
    .str.lower()
)

df['Customer Segment'] = df['Customer Segment'].replace({
    'budget': 'Budget',
    'regular': 'Regular',
    'reg': 'Regular',
    'premium': 'Premium',
    'corporate': 'Corporate',
    'corp': 'Corporate',
    'vip': 'VIP'
})

df['Customer Segment'] = df['Customer Segment'].replace(
    ['', 'nan', 'unknown', 'n/a', 'na'],
    pd.NA
)

df['Customer Segment'].value_counts(dropna=False)

Customer Segment
Budget       525
Premium      519
Regular      496
Corporate    480
VIP            5
Name: count, dtype: int64

#### 9. Fix Age column (invalid and non-numeric values)
#### Explanation: Age has numbers, blanks, 30 yrs, 25 years, 45Y, and impossible values like 150
#### or 100. Extract the number, convert to integer, and set unrealistic ages (e.g., <18 or >100) to missing.

In [9]:
df['Age'] = df['Age'].astype(str).str.extract(r'(\d+)')[0]

df['Age'] = pd.to_numeric(df['Age'], errors='coerce')

df.loc[(df['Age'] < 18) | (df['Age'] > 100), 'Age'] = pd.NA

df['Age'] = df['Age'].astype('Int64')

df['Age']

0       39
1       25
2       50
3       53
4       38
        ..
2020    52
2021    33
2022    48
2023    52
2024    23
Name: Age, Length: 2025, dtype: Int64

#### 10. Clean Room Nights column
#### Explanation: Values include numbers, N/A, 3 nights, 100, 30. Extract only the numeric part,
#### convert to integer, and decide what to do with extreme values (e.g., 100 nights).

In [10]:
df['Room Nights'] = df['Room Nights'].astype(str).str.extract(r'(\d+)')[0]

df['Room Nights'] = pd.to_numeric(
    df['Room Nights'],
    errors='coerce'
)

df.loc[df['Room Nights'] > 30, 'Room Nights'] = pd.NA

df['Room Nights'] = df['Room Nights'].astype('Int64')

df['Room Nights']

0          4
1          1
2          4
3          3
4       <NA>
        ... 
2020       3
2021       5
2022       6
2023       2
2024       5
Name: Room Nights, Length: 2025, dtype: Int64

#### 11. Fix Room Rate (mixed formats and extreme values)
#### Explanation: Room Rate contains normal numbers, 15k, $8000, 12000 INR, 100000, 250000,
#### and blanks. Convert text formats to numbers, remove currency symbols, and flag or handle unrealistic rates.

In [11]:
import numpy as np

rate = df["Room Rate"].astype("string").str.strip().str.lower()
rate = rate.str.replace(",", "", regex=False)

df["Room Rate"] = pd.to_numeric(
    rate.str.extract(r"(\d+(?:\.\d+)?)", expand=False),
    errors="coerce"
)

is_k = rate.str.contains("k", na=False)

df.loc[is_k, "Room Rate"] = (
    df.loc[is_k, "Room Rate"] * 1000
)

df.loc[df["Room Rate"] > 50000, "Room Rate"] = np.nan

df["Room Rate"]

0        7918.62
1        4822.39
2       16583.63
3        6418.64
4        5228.46
          ...   
2020     4259.26
2021     15000.0
2022    13257.76
2023     9020.55
2024     1528.93
Name: Room Rate, Length: 2025, dtype: Float64

#### 12. Clean Discount % column
#### Explanation: Discount has numbers, none, 10%, 5 %, N/A, and impossible values like 120.
#### Convert everything to a pure number between 0 and 100; treat “none” as 0 and values >100
#### as missing or capped.

In [12]:
df["Discount %"] = (
    df["Discount %"]
    .astype("string")
    .str.replace("%", "", regex=False)
    .str.strip()
)

df.loc[
    df["Discount %"].str.lower() == "none",
    "Discount %"
] = "0"

df["Discount %"] = pd.to_numeric(
    df["Discount %"],
    errors="coerce"
)

df.loc[
    (df["Discount %"] < 0) |
    (df["Discount %"] > 100),
    "Discount %"
] = np.nan

df["Discount %"].head(20)

0      6.62
1     15.25
2     23.48
3     27.67
4      32.9
5     31.11
6     34.38
7     18.39
8     15.92
9     22.24
10     34.0
11     34.5
12    24.72
13    14.09
14    15.25
15    12.33
16    13.54
17    18.81
18    17.57
19     5.39
Name: Discount %, dtype: Float64

#### 13. Handle missing and invalid Rating values
#### Explanation: Rating has numbers, blanks, 5/5, 4 stars, and values above 5 (e.g., 6, 9).
#### Convert to a 1–5 scale and set invalid entries to missing.

In [13]:
df["Rating"] = (
    df["Rating"]
    .astype("string")
    .str.extract(r"(\d+(?:\.\d+)?)", expand=False)
)

df["Rating"] = pd.to_numeric(
    df["Rating"],
    errors="coerce"
)

df.loc[
    (df["Rating"] < 1) |
    (df["Rating"] > 5),
    "Rating"
] = np.nan

df["Rating"].head(20)

0      1.4
1      2.5
2      1.2
3      2.6
4      4.2
5     <NA>
6      3.6
7      2.3
8      4.1
9      4.5
10     4.6
11     2.8
12     4.1
13     2.4
14     4.8
15     2.9
16     2.7
17     4.0
18     1.0
19     4.6
Name: Rating, dtype: Float64

#### 14. Clean Guests column (extreme values)
#### Explanation: Most values are reasonable (1–6), but some are 15, 25, or 40. Decide a sensible
#### maximum (e.g., 10) and set higher values to missing or investigate them.

In [14]:
df["Guests"] = pd.to_numeric(
    df["Guests"],
    errors="coerce"
)

df.loc[
    (df["Guests"] < 1) |
    (df["Guests"] > 10),
    "Guests"
] = np.nan

df["Guests"] = df["Guests"].astype("Int64")

df["Guests"].head(20)

0     2
1     2
2     1
3     6
4     4
5     2
6     3
7     2
8     4
9     2
10    5
11    2
12    1
13    4
14    4
15    3
16    5
17    2
18    2
19    3
Name: Guests, dtype: Int64

#### 15. Standardize Is Member column
#### Explanation: Values are mixed: 1, 0, Y, Yes, Member, blanks. Convert everything to a clear
#### Yes/No or 1/0 format.

In [15]:
df["Is Member"] = (
    df["Is Member"]
    .astype("string")
    .str.strip()
    .str.lower()
)

df["Is Member"] = df["Is Member"].map({
    "1": 1,
    "0": 0,
    "y": 1,
    "yes": 1,
    "member": 1,
    "n": 0,
    "no": 0
})

df["Is Member"] = df["Is Member"].astype("Int64")

df["Is Member"].head(20)

0     1
1     1
2     0
3     0
4     1
5     0
6     1
7     0
8     1
9     0
10    0
11    1
12    0
13    1
14    0
15    0
16    1
17    0
18    0
19    0
Name: Is Member, dtype: Int64

#### 16. Standardize Cancelled column
#### Explanation: Cancelled appears as 0, 1, Yes, Y, N. Convert to a consistent 0/1 or Yes/No format.

In [16]:
df["Cancelled"] = (
    df["Cancelled"]
    .astype("string")
    .str.strip()
    .str.lower()
)

df["Cancelled"] = df["Cancelled"].map({
    "1": 1,
    "0": 0,
    "yes": 1,
    "y": 1,
    "no": 0,
    "n": 0
})

df["Cancelled"] = df["Cancelled"].astype("Int64")

df["Cancelled"].head(20)

0     0
1     0
2     1
3     0
4     0
5     1
6     1
7     1
8     0
9     1
10    1
11    0
12    1
13    0
14    0
15    0
16    1
17    0
18    1
19    0
Name: Cancelled, dtype: Int64

#### 17. Clean Payment Method values
#### Explanation: Payment methods have inconsistent casing and spacing (Credit Card, credit
#### card, upi, Net Banking, cash , blanks). Standardize to a clean list (Credit Card, Debit Card, UPI, Net Banking, Cash).

In [17]:
df["Payment Method"] = (
    df["Payment Method"]
    .astype("string")
    .str.strip()
    .str.lower()
)

df["Payment Method"] = df["Payment Method"].map({
    "credit card": "Credit Card",
    "creditcard": "Credit Card",
    "debit card": "Debit Card",
    "debitcard": "Debit Card",
    "upi": "UPI",
    "net banking": "Net Banking",
    "netbanking": "Net Banking",
    "cash": "Cash"
})

df["Payment Method"].head(20)

0     Credit Card
1     Credit Card
2     Credit Card
3      Debit Card
4            Cash
5            Cash
6             UPI
7      Debit Card
8            Cash
9     Credit Card
10            UPI
11    Credit Card
12     Debit Card
13    Credit Card
14    Credit Card
15    Credit Card
16    Net Banking
17     Debit Card
18    Net Banking
19     Debit Card
Name: Payment Method, dtype: object

#### 18. Remove or flag duplicate Booking IDs
#### Explanation: Some Booking IDs (e.g., BKG00042, BKG01640) appear more than once. Decide
#### whether to keep the first occurrence, the most complete row, or mark them for review.

In [18]:
df["_non_missing"] = df.notna().sum(axis=1)

df = (
    df.sort_values("_non_missing", ascending=False)
      .drop_duplicates(subset=["Booking ID"], keep="first")
      .drop(columns="_non_missing")
      .reset_index(drop=True)
)

df["Booking ID"].duplicated().sum()

np.int64(0)

#### 19. Handle missing Room Type values
#### Explanation: Some rows have blank Room Type. You can either drop those rows, fill with
#### “Unknown”, or try to infer from other columns if possible.

In [19]:
df["Room Type"] = (
    df["Room Type"]
    .astype("string")
    .str.strip()
)

df["Room Type"] = df["Room Type"].fillna("Unknown")

df.loc[
    df["Room Type"].isin(["", "N/A", "NA", "Unknown"]),
    "Room Type"
] = "Unknown"

df["Room Type"].head(20)

0         Suite
1        Deluxe
2      Standard
3         Suite
4     Executive
5         Suite
6         Suite
7     Executive
8     Executive
9     Executive
10        Suite
11        Suite
12     Standard
13       Deluxe
14       Deluxe
15        Suite
16        Suite
17     Standard
18        Suite
19        Suite
Name: Room Type, dtype: string

#### 20. Fix inconsistent Final Amount calculations
#### Explanation: Final Amount should roughly equal Gross Amount − Discount Amount, but
#### some rows have missing Final Amount or clear calculation errors. Recalculate or flag inconsistencies.

In [20]:
df["Gross Amount"] = pd.to_numeric(
    df["Gross Amount"],
    errors="coerce"
)

df["Discount Amount"] = pd.to_numeric(
    df["Discount Amount"],
    errors="coerce"
)

df["Final Amount"] = pd.to_numeric(
    df["Final Amount"],
    errors="coerce"
)

expected_final = (
    df["Gross Amount"] - df["Discount Amount"]
)

df["Final Amount"] = df["Final Amount"].where(
    (df["Final Amount"] - expected_final).abs() <= 1,
    expected_final
)

df["Final Amount"] = df["Final Amount"].fillna(
    expected_final
)

df[[
    "Gross Amount",
    "Discount Amount",
    "Final Amount"
]].head(20)

,Gross Amount,Discount Amount,Final Amount
0,31674.48,2096.850576,29577.629424
1,70982.20,10725.410420,60256.789580
2,7116.81,1308.781359,5808.028641
3,54210.68,11373.400664,42837.279336
4,32166.55,6873.991735,25292.558265
5,32714.55,7403.302665,25311.247335
6,6137.45,456.012535,5681.437465
7,54435.28,4218.734200,50216.545800
8,16624.00,192.838400,16431.161600
9,82065.20,14230.105680,67835.094320


#### 21. Clean Customer Name (placeholder and invalid names)
#### Explanation: Names include Customer_1888, 12345, john doe, N/A, blanks. Decide how to
#### treat non-real names (keep as-is, replace with “Unknown”, or drop).

In [21]:
df["Customer Name"] = (
    df["Customer Name"]
    .astype("string")
    .str.strip()
)

df["Customer Name"] = df["Customer Name"].replace(
    ["", "N/A", "NA", "n/a", "None"],
    np.nan
)

df.loc[
    df["Customer Name"].str.fullmatch(r"\d+", na=False),
    "Customer Name"
] = "Unknown"

df["Customer Name"] = df["Customer Name"].fillna("Unknown")

df["Customer Name"].head(20)

0     Customer_1888
1     Customer_1608
2     Customer_1356
3     Customer_1242
4     Customer_0184
5     Customer_1854
6     Customer_0957
7     Customer_0002
8     Customer_1733
9     Customer_1096
10    Customer_1330
11          unknown
12    Customer_0700
13    Customer_1626
14    Customer_1460
15    Customer_1337
16    Customer_0001
17    Customer_0134
18    Customer_0851
19    Customer_0131
Name: Customer Name, dtype: string

#### 22. Convert all numeric columns to proper numeric data types
#### Explanation: Many columns that should be numbers are currently stored as text because of
#### mixed formats. Convert them to integer or float so calculations work correctly.

In [22]:
numeric_columns = [
    "Age",
    "Room Nights",
    "Room Rate",
    "Discount %",
    "Rating",
    "Guests",
    "Gross Amount",
    "Discount Amount",
    "Final Amount",
    "Profit"
]

for col in numeric_columns:
    if col in df.columns:
        df[col] = pd.to_numeric(
            df[col],
            errors="coerce"
        )

df["Age"] = df["Age"].astype("Int64")
df["Room Nights"] = df["Room Nights"].astype("Int64")
df["Guests"] = df["Guests"].astype("Int64")

print(df[numeric_columns].dtypes)

Age                  Int64
Room Nights          Int64
Room Rate          Float64
Discount %         Float64
Rating             Float64
Guests               Int64
Gross Amount       float64
Discount Amount    float64
Final Amount       float64
Profit             float64
dtype: object


#### 23. Handle missing values in key financial columns (Gross Amount, Discount Amount, Final
#### Amount, Profit)
#### Explanation: Some financial columns have blanks. Decide whether to drop those rows, fill
#### with 0, or calculate missing values from other columns when possible.

In [23]:
financial_columns = [
    "Gross Amount",
    "Discount Amount",
    "Final Amount",
    "Profit"
]

for col in financial_columns:
    df[col] = pd.to_numeric(df[col], errors="coerce")

df["Gross Amount"] = df["Gross Amount"].fillna(0)

df["Discount Amount"] = df["Discount Amount"].fillna(0)

df["Final Amount"] = df["Final Amount"].fillna(
    df["Gross Amount"] - df["Discount Amount"]
)

df["Profit"] = df["Profit"].fillna(
    df["Final Amount"]
)

df[financial_columns].head(20)

,Gross Amount,Discount Amount,Final Amount,Profit
0,31674.48,2096.850576,29577.629424,12676.243171
1,70982.20,10725.410420,60256.789580,17640.895377
2,7116.81,1308.781359,5808.028641,1203.368368
3,54210.68,11373.400664,42837.279336,12944.492212
4,32166.55,6873.991735,25292.558265,6013.058253
5,32714.55,7403.302665,25311.247335,8887.453503
6,6137.45,456.012535,5681.437465,2604.046163
7,54435.28,4218.734200,50216.545800,21426.158285
8,16624.00,192.838400,16431.161600,3989.789322
9,82065.20,14230.105680,67835.094320,18447.136960


#### 24. Remove rows with completely invalid dates
#### Explanation: Dates like 32/13/2025 are impossible. Identify and either remove those rows or set the date to missing.

In [24]:
df["Booking Date"] = pd.to_datetime(
    df["Booking Date"],
    errors="coerce"
)

df["Booking Date"] = df["Booking Date"].dt.strftime(
    "%Y-%m-%d"
)

df["Booking Date"].head(20)

0     2025-05-17
1     2025-06-21
2     2025-09-11
3     2025-02-06
4     2025-08-11
5     2025-04-02
6     2025-06-24
7     2025-10-10
8     2025-01-02
9     2025-08-13
10    2025-01-13
11    2025-08-13
12    2025-07-04
13    2025-06-09
14    2025-01-28
15    2025-10-07
16    2025-09-15
17    2025-08-13
18    2025-02-01
19    2025-12-29
Name: Booking Date, dtype: object

#### 25. Standardize boolean-like columns (Is Member and Cancelled) to the same format
#### Explanation: Keeping both columns in the same style (both 0/1 or both Yes/No) makes
#### filtering and analysis much easier and less error-prone.

In [25]:
df["Is Member"] = (
    df["Is Member"]
    .astype("string")
    .str.strip()
    .str.lower()
    .map({
        "1": 1,
        "0": 0,
        "yes": 1,
        "y": 1,
        "member": 1,
        "no": 0,
        "n": 0
    })
)

df["Cancelled"] = (
    df["Cancelled"]
    .astype("string")
    .str.strip()
    .str.lower()
    .map({
        "1": 1,
        "0": 0,
        "yes": 1,
        "y": 1,
        "no": 0,
        "n": 0
    })
)

df["Is Member"] = df["Is Member"].astype("Int64")
df["Cancelled"] = df["Cancelled"].astype("Int64")

print(df[["Is Member", "Cancelled"]].head(20))

    Is Member  Cancelled
0           1          0
1           0          1
2           1          0
3           0          0
4           0          1
5           0          0
6           0          1
7           1          0
8           1          0
9           1          0
10          1          1
11          1          0
12          1          1
13          1          1
14          0          0
15          1          1
16          0          1
17          1          0
18          0          0
19          1          1


#### 26. Trim extra spaces from all text columns
#### Explanation: Many text fields have leading/trailing spaces (e.g., "Budget ", "Mumbai ").
#### Removing them prevents “Budget” and “Budget ” from being treated as different categories.

In [26]:
text_columns = df.select_dtypes(
    include=["object", "string"]
).columns

for col in text_columns:
    df[col] = df[col].str.strip()

print(df[text_columns].head(20))

   Booking ID  Customer Name Booking Date       City  Room Type  \
0    BKG01888  Customer_1888   2025-05-17     Mumbai      Suite   
1    BKG01608  Customer_1608   2025-06-21     Mumbai     Deluxe   
2    BKG01356  Customer_1356   2025-09-11    Unknown   Standard   
3    BKG01242  Customer_1242   2025-02-06  Hyderabad      Suite   
4    BKG00184  Customer_0184   2025-08-11  Bengaluru  Executive   
5    BKG01854  Customer_1854   2025-04-02    Unknown      Suite   
6    BKG00957  Customer_0957   2025-06-24  Bengaluru      Suite   
7    BKG00002  Customer_0002   2025-10-10  Bengaluru  Executive   
8    BKG01733  Customer_1733   2025-01-02    Unknown  Executive   
9    BKG01096  Customer_1096   2025-08-13  Hyderabad  Executive   
10   BKG01330  Customer_1330   2025-01-13    Unknown      Suite   
11   BKG01398        unknown   2025-08-13     Mumbai      Suite   
12   BKG00700  Customer_0700   2025-07-04     Mumbai   Standard   
13   BKG01626  Customer_1626   2025-06-09  Bengaluru     Delux

#### 27. Fix extreme or impossible Age values
#### Explanation: Ages such as 150, 120, or 100 are unrealistic for hotel customers. Set a
#### reasonable range (e.g., 18–90) and replace values outside it with missing.

In [27]:
df["Age"] = pd.to_numeric(
    df["Age"],
    errors="coerce"
)

df.loc[
    (df["Age"] < 18) |
    (df["Age"] > 90),
    "Age"
] = np.nan

df["Age"] = df["Age"].astype("Int64")

print(df["Age"].head(20))

0     39
1     62
2     41
3     20
4     54
5     37
6     62
7     32
8     50
9     66
10    43
11    22
12    24
13    67
14    34
15    48
16    35
17    38
18    64
19    38
Name: Age, dtype: Int64


#### 28. Handle missing Gender values
#### Explanation: Some rows have blank Gender. You can fill with “Unknown”, drop the rows, or
#### leave them as missing depending on your analysis needs.

In [28]:
df["Gender"] = (
    df["Gender"]
    .astype("string")
    .str.strip()
)

df["Gender"] = df["Gender"].fillna("Unknown")

df.loc[
    df["Gender"].isin(["", "N/A", "NA"]),
    "Gender"
] = "Unknown"

print(df["Gender"].value_counts())

Gender
Female    988
Male      943
Other      69
Name: count, dtype: Int64


#### 29. Correct obvious typos in categorical columns
#### Explanation: Examples include delux → Deluxe, exec → Executive, Hydrabad → Hyderabad,
#### Bangalore → Bengaluru. Fixing these improves grouping and reporting accuracy.

In [29]:
df["Room Type"] = df["Room Type"].replace({
    "delux": "Deluxe",
    "exec": "Executive"
})

df["City"] = df["City"].replace({
    "Hydrabad": "Hyderabad",
    "Bangalore": "Bengaluru"
})

print(df[["Room Type", "City"]].head(20))

    Room Type       City
0       Suite     Mumbai
1      Deluxe     Mumbai
2    Standard    Unknown
3       Suite  Hyderabad
4   Executive  Bengaluru
5       Suite    Unknown
6       Suite  Bengaluru
7   Executive  Bengaluru
8   Executive    Unknown
9   Executive  Hyderabad
10      Suite    Unknown
11      Suite     Mumbai
12   Standard     Mumbai
13     Deluxe  Bengaluru
14     Deluxe    Unknown
15      Suite     Mumbai
16      Suite     Mumbai
17   Standard    Unknown
18      Suite    Unknown
19      Suite    Unknown


#### 30. Decide how to treat rows with too many missing values
#### Explanation: Some rows are missing Customer Name, City, Room Type, Age, and other
#### important fields at the same time. Decide a threshold (e.g., more than 40% missing) and drop or flag those incomplete rows.

In [30]:
missing_percentage = df.isna().mean(axis=1)

df = df[
    missing_percentage <= 0.40
].reset_index(drop=True)

print(df.shape)

(2000, 22)
